### PR TITLE
Issue 221 - ability to inject topics into tests

### PR DESCRIPTION
**Pull request by k-wall**
### Type of change

_Select the type of your PR_

- Enhancement / new feature

### Description

Ability to inject test topics into tests. 

The user signals their wish for a topic to be injected by the use of a `Topic` type parameter.  On encountering this, the Junit extension will create a topic on the broker using a random name, and let the test know the name via the `Topic` parameter.    The test is guaranteed that the topic will exist before the test method commences.

Passing topic config, replication factor and number of partitions is supported.

### Additional Context

_Why are you making this pull request?_

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [x] Write tests
- [x] Make sure all tests pass
- [x] Update documentation
- [x] Reference relevant issue(s) and close them after merging
- [x] Update CHANGELOG.md

### CODE REVIEW

File: junit5-extension/src/main/java/io/kroxylicious/testing/kafka/junit5ext/TopicPartitions.java
Comment (SamBarker) [inline junit5-extension/src/main/java/io/kroxylicious/testing/kafka/junit5ext/TopicPartitions.java:22]: I don't have strong views on this but I'd like to hear your thoughts on why partition and replica counts are separate annotations rather than properties of the topic annotation. 

The later feels like a slightly more natural expression to me (probably using nullable fields for inheriting broker defaults)

File: impl/src/main/java/io/kroxylicious/testing/kafka/internal/AnonymousAdminSource.java
Comment (SamBarker) [inline impl/src/main/java/io/kroxylicious/testing/kafka/internal/AnonymousAdminSource.java:11]: We can always refactor later but why is "Anonymous" part of the API? I see that as, an important, implementation detail. 

To expand on that we could supply a an admin client with fixed credentials which we have also injected into the cluster and satisfy the contract we need from this interface.

**Review by SamBarker — APPROVED** [review]
I'm happy in principal with this change. 

I do have a few questions about the exact shape, but they could be equally well addressed in follow up PRs as being fixed before this one merges.

**Review by franvila — APPROVED** [review]

File: junit5-extension/src/main/java/io/kroxylicious/testing/kafka/junit5ext/TopicPartitions.java
Comment (k-wall) [reply junit5-extension/src/main/java/io/kroxylicious/testing/kafka/junit5ext/TopicPartitions.java:22]: > I don't have strong views on this but I'd like to hear your thoughts on why partition and replica counts are separate annotations rather than properties of the topic annotation.

My motivation was the cardinality - TopicConfig naturally needs to be repeatable in order to support > 1.

Comment (SamBarker) [reply junit5-extension/src/main/java/io/kroxylicious/testing/kafka/junit5ext/TopicPartitions.java:22]: A repeatable topicConfig makes sense but i meant  partitions and replication factor specifically

File: impl/src/main/java/io/kroxylicious/testing/kafka/internal/AnonymousAdminSource.java
Comment (k-wall) [reply impl/src/main/java/io/kroxylicious/testing/kafka/internal/AnonymousAdminSource.java:11]: Good point, it is an implementation detail.

File: junit5-extension/src/main/java/io/kroxylicious/testing/kafka/junit5ext/KafkaClusterExtension.java
Comment (k-wall) [reply junit5-extension/src/main/java/io/kroxylicious/testing/kafka/junit5ext/KafkaClusterExtension.java]: Done

File: junit5-extension/src/main/java/io/kroxylicious/testing/kafka/junit5ext/TopicPartitions.java
Comment (k-wall) [reply junit5-extension/src/main/java/io/kroxylicious/testing/kafka/junit5ext/TopicPartitions.java:22]: Ah, so you mean introduce a `@Topic` annotation which would carry the replicationFactor/replicationFactor fields.  I presume that actually adding the annotation would remain optional.

So it would look like:

```
@Topic(replicationFactor = 2) Topic myTopic 
```

this form would work too.

```
Topic myTopic 
```

File: junit5-extension/src/main/java/io/kroxylicious/testing/kafka/junit5ext/KafkaClusterExtension.java
Comment (k-wall) [reply junit5-extension/src/main/java/io/kroxylicious/testing/kafka/junit5ext/KafkaClusterExtension.java]: Done. I made it package private.

File: junit5-extension/src/main/java/io/kroxylicious/testing/kafka/junit5ext/TopicPartitions.java
Comment (k-wall) [reply junit5-extension/src/main/java/io/kroxylicious/testing/kafka/junit5ext/TopicPartitions.java:22]: Alternatively,  we could chose to special case the config names `replication.factor` and `partitions`, and use those to pass the configuration.  Those names aren't understood by the Kafka Client, but they do have currency already in Kafka Connect (`org.apache.kafka.connect.runtime.TopicCreationConfig#REPLICATION_FACTOR_CONFIG`).  That would avoid the need for the additional annotation but would mean we have the extra javadoc.

**Comment by k-wall** [comment]
@SamBarker @franvila  I want to point out the package/module location of the new @TopicConfig annotation.  See #234 for more details.

File: junit5-extension/src/main/java/io/kroxylicious/testing/kafka/junit5ext/Topic.java
Comment (k-wall) [inline junit5-extension/src/main/java/io/kroxylicious/testing/kafka/junit5ext/Topic.java:17]: I did think about exposing the topicUuid here.  I decided, YAGNI.

Comment (SamBarker) [reply junit5-extension/src/main/java/io/kroxylicious/testing/kafka/junit5ext/Topic.java:17]: Yeah, I think YAGNI is the right call.

It will be very cheap to add if we come up with a motivation for exposing both so lets leave it out for now.

File: junit5-extension/src/main/java/io/kroxylicious/testing/kafka/junit5ext/TopicPartitions.java
Comment (SamBarker) [reply junit5-extension/src/main/java/io/kroxylicious/testing/kafka/junit5ext/TopicPartitions.java:22]: > `@Topic(replicationFactor = 2) Topic myTopic `

Was exactly what I had in mind. I confess `@Topic Topic` does look a little odd. 

I think your right about `@TopicParitions` vs using `@TopicConfig` and handling repetition so the dedicated annotations make more sense than re-using `TopicConfig`.

*My personal* preference would still be an optional `@Topic` but I'm happy to merge this and come revisit later.

Comment (k-wall) [reply junit5-extension/src/main/java/io/kroxylicious/testing/kafka/junit5ext/TopicPartitions.java:22]: I have a marginal preference for `@TopicParitions/`@TopicReplicationFactor` over `@Topic Topic` so I'm going to stick with my way.

**Comment by k-wall** [comment]
> @SamBarker @franvila I want to point out the package/module location of the new @TopicConfig annotation. See #234 for more details.

PR is good to merge although could be affected by the outcome of #234.  Holding off merge until that reaches conclusion.

**Pull request merged**
